CPU starvation checker

TimWSpence opened the pull request:
Run a background fiber that detects CPU starvation by looking for clock drift when sleeping. Writes a warning to stderr if starvation is detected

It can be configured via the new system properties
- `cats.effect.cpu.starvation.check.interval: FiniteDuration`: how long it should sleep for on each loop
- `cats.effect.cpu.starvation.check.initialDelay: Duration`: how long to wait before starting the check (set to `Duration.Inf` to disable entirely)
- `cats.effect.cpu.starvation.check.threshold: Double`: how much clock drift can be tolerated before warning. The max drift tolerated is `(1 + threshold) * interval`


Resolves https://github.com/typelevel/cats-effect/issues/2762

TimWSpence committed:
CPU starvation checker

TimWSpence committed:
Configurable cpu starvation check

TimWSpence committed:
Configure cpu starvation check via system properties

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/IORuntimeConfigCompanionPlatform.scala:
```diff
@@ -44,11 +44,33 @@ private[unsafe] abstract class IORuntimeConfigCompanionPlatform { this: IORuntim
+    val cpuStarvationCheckInterval =
+      Try(System.getProperty("cats.effect.cpu.starvation.check.interval"))
+        .map(Duration(_))
+        .flatMap { d => Try(d.asInstanceOf[FiniteDuration]) }
```
I think you can just do `case fd: FiniteDuration => ...` here, instead of ... this :P

armanbilge replied:
It might help to transform the `Try` to an `Option`.

TimWSpence replied:
haha yup, either of those would be preferable 😂

TimWSpence committed:
Add arguments to copy

TimWSpence committed:
Make starvation check initial delay a Duration

So it can be effectively disabled by passing `Duration.Inf`

TimWSpence commented:
@armanbilge are you still planning to do https://github.com/typelevel/cats-effect/pull/3098? Or should I add another parameter to toggle this on/off (instead of allowing `initialDelay = Duration.Inf`)?

armanbilge commented:
Yup, feel free to base off https://github.com/typelevel/cats-effect/pull/3098.

djspiewak commented on core/jvm/src/main/scala/cats/effect/unsafe/IORuntimeConfigCompanionPlatform.scala:
```diff
@@ -44,11 +44,33 @@ private[unsafe] abstract class IORuntimeConfigCompanionPlatform { this: IORuntim
+        .getOrElse(DefaultCpuStarvationCheckInterval)
+
+    val cpuStarvationCheckInitialDelay =
+      Try(System.getProperty("cats.effect.cpu.starvation.check.initial.delay"))
```
I'm not sure I like the paradigm of splitting up words in property names with `.`. :-) Something like `cats.effect.starvation.check.initialDelay` makes a bit more sense in terms of the hierarchy.

TimWSpence replied:
sure! And no it doesn't have to be based on top of that at all - I was just trying to avoid an additional parameter to disable it entirely. But I can rebase and do that - would be nice to unblock this!

armanbilge replied:
https://github.com/typelevel/cats-effect/pull/3098 already merged in so this point is moot :)

TimWSpence replied:
Haha great!!

djspiewak reviewed (changes requested):
Does this *have* to build on top of the `FiniteDuration` => `Duration` PR? That has some subtleties so I wanted to untangle these things ideally.

djspiewak commented:
Oh additional question: can we do this in a common location so it can be shared across the `IOApp`s? This is even more useful on JavaScript than on the JVM!

TimWSpence commented:
> Oh additional question: can we do this in a common location so it can be shared across the `IOApp`s? This is even more useful on JavaScript than on the JVM!

Yes, good point!

TimWSpence committed:
Merge remote-tracking branch 'upstream/series/3.x' into cpu-starvation-checker

TimWSpence committed:
Don't split property name with .

TimWSpence committed:
Run cpu starvation check for all platforms

armanbilge commented on core/shared/src/main/scala/cats/effect/CpuStarvationCheck.scala:
```diff
@@ -0,0 +1,22 @@
+          .map(_ - now)
+          .flatMap { delta =>
+            Console[IO]
+              .errorln("[WARNING] you're CPU threadpool is probably starving")
```
😛 
```suggestion
              .errorln("[WARNING] your CPU threadpool is probably starving")
```

TimWSpence replied:
Already spotted it! 😛

TimWSpence committed:
Create headers

TimWSpence committed:
Allow the initialDelay property to be infinite

TimWSpence committed:
Absolute import for sake of scalafix

TimWSpence committed:
ARGH scalafmt

TimWSpence committed:
Test for cpu starvation check

TimWSpence committed:
Actually calculate fibonacci

djspiewak commented on tests/shared/src/main/scala/catseffect/examples.scala:
```diff
@@ -107,4 +107,15 @@ package examples {
+      case n => fib(n - 1) + fib(n - 2)
+    }
+
+    val run = IO.delay(fib(1000)).replicateA_(50)
```
Fwiw, `Thread.sleep` also works :-P

TimWSpence replied:
haha good point!

djspiewak commented on core/shared/src/main/scala/cats/effect/CpuStarvationCheck.scala:
```diff
@@ -0,0 +1,38 @@
+private[effect] object CpuStarvationCheck {
+  def run(runtimeConfig: IORuntimeConfig): IO[Unit] =
+    IO.monotonic
+      .flatMap { now =>
+        IO.sleep(runtimeConfig.cpuStarvationCheckInterval) >> IO
+          .monotonic
+          .map(_ - now)
+          .flatMap { delta =>
+            Console[IO]
+              .errorln("[WARNING] your CPU threadpool is probably starving")
+              .whenA(delta > runtimeConfig.cpuStarvationCheckThreshold)
```
Couple thoughts on this…

First, I think we should probably have the starvation threshold be a *multiple* of the interval. In other words, the check threshold should be a `Double` like `0.1` or `0.2` (representing 10% of 20%, respectively). Then the calculation itself would be something like this:

```suggestion
      .flatMap { now =>
        IO.sleep(runtimeConfig.cpuStarvationCheckInterval) >> IO
          .monotonic
          .map(_ - now)
          .flatMap { delta =>
            Console[IO]
              .errorln("[WARNING] Your CPU is probably starving. Consider reducing the granularity of your `delay`s or adding more `cede`s. This may also be a sign that you are unintentionally running blocking I/O operations (such as File or InetAddress) without the `blocking` combinator.")
              .whenA(delta >= runtimeConfig.cpuStarvationCheckInterval.toDouble * (1 + runtimeConfig.cpuStarvationCheckThreshold))
```

My second thought is that we should be more explanatory in the error message. I took a stab at it in the above. Probably needs line wrapping at the very least.

TimWSpence replied:
Yep, thank you!

djspiewak commented on core/shared/src/main/scala/cats/effect/unsafe/IORuntimeConfig.scala:
```diff
@@ -93,6 +119,9 @@ object IORuntimeConfig extends IORuntimeConfigCompanionPlatform {
   private[unsafe] def DefaultTraceBufferSize = 16
   private[unsafe] def DefaultShutdownHookTimeout = Duration.Inf
+  private[unsafe] def DefaultCpuStarvationCheckInterval = 1.second
+  private[unsafe] def DefaultCpuStarvationCheckInitialDelay = 10.millis
```
Is `10.millis` a sufficient initial delay? That's usually well within warmup times. I would think more like `10.seconds` would be better.

TimWSpence replied:
Yeah this was an entirely arbitrary value I stuck in whilst getting things to compile 😅 `10.seconds` sounds reasonable 👍

djspiewak commented on core/shared/src/main/scala/cats/effect/unsafe/IORuntimeConfig.scala:
```diff
@@ -93,6 +119,9 @@ object IORuntimeConfig extends IORuntimeConfigCompanionPlatform {
   private[unsafe] def DefaultShutdownHookTimeout = Duration.Inf
+  private[unsafe] def DefaultCpuStarvationCheckInterval = 1.second
+  private[unsafe] def DefaultCpuStarvationCheckInitialDelay = 10.millis
+  private[unsafe] def DefaultCpuStarvationCheckThreshold = 100.millis
```
If we accept my suggestion from above, this would become `0.1d`

djspiewak commented on tests/shared/src/main/scala/catseffect/examples.scala:
```diff
@@ -61,19 +62,18 @@ package examples {
-
     def run(args: List[String]): IO[ExitCode] =
-      IO.pure(ExitCode.Success)
+      IO.println("boom").whenA(!r.eq(runtime)) >> IO.pure(ExitCode.Success)
```
Why did this change?

TimWSpence replied:
Sorry, I hadn't finished getting this to work properly. But the problem is that we are now using the config for something other than instantiating the runtime so the way the test previously worked (`override def runtimeConfig = sys.error("boom")`) has to change

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/IORuntimeConfigCompanionPlatform.scala:
```diff
@@ -44,11 +44,32 @@ private[unsafe] abstract class IORuntimeConfigCompanionPlatform { this: IORuntim
         .map(Duration(_))
         .getOrElse(DefaultShutdownHookTimeout)
 
+    val cpuStarvationCheckInterval =
+      Try(System.getProperty("cats.effect.cpu.starvation.check.interval"))
+        .map(Duration(_))
+        .flatMap { d => Try(d.asInstanceOf[FiniteDuration]) }
+        .getOrElse(DefaultCpuStarvationCheckInterval)
+
+    val cpuStarvationCheckInitialDelay =
+      Try(System.getProperty("cats.effect.cpu.starvation.check.initialDelay"))
+        .map(Duration(_))
+        .getOrElse(DefaultCpuStarvationCheckInitialDelay)
+
+    val cpuStarvationCheckThreshold =
+      Try(System.getProperty("cats.effect.cpu.starvation.check.threshold"))
+        .map(Duration(_))
+        .flatMap { d => Try(d.asInstanceOf[FiniteDuration]) }
+        .getOrElse(DefaultCpuStarvationCheckThreshold)
```
We need to add these for JS and Native too 😅

djspiewak committed:
Avoid spurious wake-up in `Concurrent` queues at capacity

djspiewak committed:
Merge pull request #3180 from djspiewak/bug/synchronous-fifo

Avoid spurious wake-up in `Concurrent` queues at capacity

typelevel-steward[bot] committed:
Update sbt-mdoc to 2.3.6 in series/3.x

djspiewak committed:
Merge pull request #3188 from typelevel/update/series/3.x/sbt-mdoc-2.3.6

Update sbt-mdoc to 2.3.6 in series/3.x

TimWSpence commented:
Apologies @djspiewak and @armanbilge for being so slow on this. Will try to get it over the line this week

TimWSpence committed:
Sleep instead of fibonacci for thread starvation

TimWSpence committed:
Explicitly set cpu starvation check config in test

TimWSpence committed:
Fix assertion

TimWSpence committed:
Make cpu starvation threshold a double

TimWSpence committed:
Only check monotonic time once per loop

TimWSpence committed:
Improve error message

TimWSpence committed:
JS env overrides for cpu starvation check config

TimWSpence committed:
Native env overrides for cpu starvation check config

TimWSpence committed:
Merge remote-tracking branch 'upstream/series/3.x' into cpu-starvation-checker

TimWSpence committed:
Fix the merge :sob:

TimWSpence committed:
Scalafix

TimWSpence committed:
Get CPU starvation test to terminate in the hope of more reliability

TimWSpence committed:
ARGH SCALAFMT

TimWSpence committed:
Platform-specific starvation checker tests

TimWSpence committed:
scalafix again

TimWSpence committed:
Shared IOApp test again

- Register it to make ScalaJS happy
- Back to Fibonacci as Thread.sleep doesn't work on scalaJS
- Don't use Runtime for cross-building reasons
- Fewer fibers to try to strike a balance where it should fail on JVMs
  that don't have a ludicrous number of cpus available but also
  completes within my lifetime on JS

TimWSpence committed:
s/reducing/increasing/

TimWSpence committed:
Validate that cpu starvation check threshold is > 0

TimWSpence committed:
Line wrapping for cpu starvation warning message

djspiewak commented on core/js/src/main/scala/cats/effect/unsafe/IORuntimeConfigCompanionPlatform.scala:
```diff
@@ -53,12 +53,35 @@ private[unsafe] abstract class IORuntimeConfigCompanionPlatform { this: IORuntim
 
+    val cpuStarvationCheckInterval =
+      process
+        .env("CATS_EFFECT_CPU_STARVATION_CHECK_INTERNAL")
```
Typo!

```suggestion
        .env("CATS_EFFECT_CPU_STARVATION_CHECK_INTERVAL")
```

TimWSpence replied:
ouch! Sorry about that!!

djspiewak commented on core/js/src/main/scala/cats/effect/unsafe/IORuntimeConfigCompanionPlatform.scala:
```diff
@@ -53,12 +53,35 @@ private[unsafe] abstract class IORuntimeConfigCompanionPlatform { this: IORuntim
+
+    val cpuStarvationCheckInitialDelay =
+      process
+        .env("CATS_EFFECT_CPU_STARVATION_CHECK_INITIALDELAY")
```
```suggestion
        .env("CATS_EFFECT_CPU_STARVATION_CHECK_INITIAL_DELAY")
```

djspiewak commented on core/native/src/main/scala/cats/effect/unsafe/IORuntimeConfigCompanionPlatform.scala:
```diff
@@ -50,12 +50,32 @@ private[unsafe] abstract class IORuntimeConfigCompanionPlatform { this: IORuntim
         .getOrElse(DefaultReportUnhandledFiberErrors)
 
+    val cpuStarvationCheckInterval =
+      Option(System.getenv("CATS_EFFECT_CPU_STARVATION_CHECK_INTERNAL"))
```
```suggestion
      Option(System.getenv("CATS_EFFECT_CPU_STARVATION_CHECK_INTERVAL"))
```

djspiewak commented on core/native/src/main/scala/cats/effect/unsafe/IORuntimeConfigCompanionPlatform.scala:
```diff
@@ -50,12 +50,32 @@ private[unsafe] abstract class IORuntimeConfigCompanionPlatform { this: IORuntim
+        .getOrElse(DefaultCpuStarvationCheckInterval)
+
+    val cpuStarvationCheckInitialDelay =
+      Option(System.getenv("CATS_EFFECT_CPU_STARVATION_CHECK_INITIALDELAY"))
```
```suggestion
      Option(System.getenv("CATS_EFFECT_CPU_STARVATION_CHECK_INITIAL_DELAY"))
```

This makes me feel like we should find a way to share some of this between JS and Scala Native. Probably not for this PR but something to think about for a follow-up.

TimWSpence replied:
Yes, would be nice!

djspiewak commented on core/shared/src/main/scala/cats/effect/CpuStarvationCheck.scala:
```diff
@@ -0,0 +1,47 @@
+    """[WARNING] Your CPU is probably starving. Consider increasing the granularity
+      |of your `delay`s or adding more `cede`s. This may also be a sign that you are
+      |unintentionally running blocking I/O operations (such as File or InetAddress)
+      |without the `blocking` combinator.""".stripMargin
```
```suggestion
    """[WARNING] Your CPU is probably starving. Consider increasing the granularity
      |of your delays or adding more cedes. This may also be a sign that you are
      |unintentionally running blocking I/O operations (such as File or InetAddress)
      |without the blocking combinator.""".stripMargin
```

Had a look at `TestControl` and it seems that we skip the backticks usually.

djspiewak reviewed (changes requested):
Just need a couple adjustments and then I think we're good to go. I changed the base so we're now pointing to 3.4.x

TimWSpence committed:
Typos and remove backticks from warning

TimWSpence commented:
> Just need a couple adjustments and then I think we're good to go. I changed the base so we're now pointing to 3.4.x

Thank you!!

djspiewak reviewed (approved):
As another follow-up, we should document the new envars and properties.
